[Frontend] Jax dynamic API support in quantum control-flow primitives

grwlf opened the pull request:
In this PR we enable the Jax dynamic API in the Catalyst quantum control flow primitives: while-loops, for-loops, conditionals in both quantum and classical tracing modes.

An extended description of the solution is available in the [readme](https://github.com/PennyLaneAI/catalyst/blob/dynshape-quantum-primitives/FRONTEND.md).

Earlier related internal documents:
- [Solution comparison](https://www.notion.so/xanaduai/Answers-for-Josh-23a1cdaec47248e08d436d61f9e2526c#373ab50fc73947c9a7a688a91e09fa29)
- [Jax dynamic API 101](https://www.notion.so/xanaduai/Jax-dynamic-API-vs-Catalyst-primitives-88f5b2e5bcb5458baa1df88a022956c3)
- [Jax communication topics](https://www.notion.so/xanaduai/Jax-communication-topics-02807cdb9f30445caed3f8bb3b144a3f#8525429745e644de9a79401dd698122e)

Tasks:

* [x] conditionals
* [x] while-loops
* [x] for-loops
* [ ] Finalization and clean-up 
  * [x] Fix remaining tests
  * [x] Fix CodeFactor issues
  * [x] Merge the PR with the recent main
  * [ ] Write the documentation and clean-up
    * [x] Write a Readme file describing the proposed tracing architecture.
    * [x] Split the `utils/jax_extras.py` into `jax_extras/tracing.py`, `jax_extras/lowering.py` and `jax_extras/patches.py`.
    * [ ] Code refactoring (Sergei expects David to put ToDo marks highlighting the problems)
      - Find better names for some classes/variables
      - Write comments where required
  * [x] Fix CodeCov issues

[sc-50789]

dime10 commented on frontend/catalyst/jax_primitives.py:
```diff
@@ -1195,26 +1224,41 @@ def emit_branches(preds, branch_jaxprs, ip):
 
+@while_p.def_abstract_eval
+def _while_loop_abstract_eval(*in_type, body_jaxpr, nimplicit, preserve_dimensions, **kwargs):
+    _assert_jaxpr_without_constants(body_jaxpr)
```
What is the implication of this restriction? Do you have an example?

grwlf replied:
For me this is more a self-documenting code. Literally, body jaxpr of a loop should not have any constants, all constants should be moved to arguments using `convert_constvars_jaxpr`, so we do it in our new generic tracing function [here](https://github.com/PennyLaneAI/catalyst/blob/f03253d8e5ffeee5b91ba79514057d8d813bc874/frontend/catalyst/utils/jax_extras.py#L385)

dime10 replied:
Sure the function name makes it very clear what is being asserted, what is not clear is why. Did we assert this previously? If not, what changed? What is the resulting effect of asserting this?

This is what I wanted to find out with my comment.

grwlf replied:
We did not assert this previously, nor did we rely on this fact here. With dynamic dimensions, we need to know the number of constants to process type links correctly. So I suggest to add this assertion of having no constants at this point.

dime10 replied:
Am I understanding you correctly that we are introducing a regression that prevents a user from accessing variables inside the body of while loop that are not passed in via the arguments? For example, this would no longer be allowed:

```py
x = 5

@while_loop(lambda ...)
def f(y):
  new_y = x * y
  return new_y
```

grwlf replied:
I think it would be correct to say that we are not going to use arrays with dynamic shapes in this mode. Operations on the statically-shaped arrays should be supported, so I thought about it as of limitation, not a regression.

dime10 replied:
So you are saying that variable capture (which jax calls constants) is only prevented when the dynamic array feature is active? 

Unless I'm missing something, this assertion is applied to all while loop primitives.

grwlf replied:
> So you are saying that variable capture (which jax calls constants) is only prevented when the dynamic array feature is active?
Yes, I think so. Look [here](https://github.com/PennyLaneAI/catalyst/blob/2cf433c2aa009feccb7e63213ffb8bfaaf541558/frontend/catalyst/utils/jax_extras.py#L385) - we move constants to variables at the time of `out_sig.out_jaxpr()` [here](https://github.com/PennyLaneAI/catalyst/blob/2cf433c2aa009feccb7e63213ffb8bfaaf541558/frontend/catalyst/pennylane_extensions.py#L1953)

dime10 replied:
Looks like you are right, tests with outer variable capture seem to run just fine.

In that case I don't understand what input would trip up the assertion. Do you have an example?

grwlf replied:
A failing example would be a `body_jaxpr` for which someone forgot to convert constants into inputs. Moving constants into inputs means that one takes the responsibility for remembering the constant values somewhere outside the Jaxpr data. We need to do this because constants need to be lowered too. I think it is correct to say that this assertion checks that we indeed do this.

dime10 replied:
Ok so I tried digging into this concept of jaxpr constants (`constvars`) to figure out when they are actually used. The page on "understanding jaxpr" was helpful in this regard: https://jax.readthedocs.io/en/latest/jaxpr.html#constant-vars
Some insights:
- `constvars` are only used for constant (i.e. non-tracer) _array_ values, that is scalars are embedded directly into the jaxpr
- `constvars` are used for values that appear inside a function which are _hoisted out_ as jaxpr arguments, outside variable capture has nothing to with it
- `constvars` are hoisted out all the way to the top-level jaxpr, so they shouldn't appear in higher order primitives (primitives with their own jaxpr)

Does this match your understanding? Did you find, manipulate, or otherwise interact with jaxpr constants while working on this feature? From my testing I was only ever able to generate constants in the top-level jaxpr anyways (so the jaxpr for the qjit function, never in primitive/nested jaxprs).

dime10 replied:
> I would rather say that constants are not used for literal scalars, but might be used for other values. The to_jaxpr2 tracer does return outer tracers as constants of the inner (body) program. See the [related function](https://github.com/google/jax/blob/35071c74be259b8018b46ed4aafd624b179586e4/jax/_src/interpreters/partial_eval.py#L1808)

That's what I said no? (constant) scalar values never end up as `constvars`, because they are embedded directly into the IR as literals. 

> I believe that we do not use this mechanism for our primitives. Compare with the code for Jax' own primitives [here](https://github.com/google/jax/blob/35071c74be259b8018b46ed4aafd624b179586e4/jax/_src/lax/lax.py#L2422) where they do use some constants folding machinery.

The discussion is about constant hoisting, I don't see how folding is related. 

> I think that Jax does not treat our primitives as higher order ones. This is because we pass jaxprs as keyword args without letting Jax know about it in any way. So, I believe that all our body jaxprs are top-level in this sense, and we are responsible for dealing with constants at every layer.

I have seen no evidence of this. I have tested examples where variables are hoisted out to the top-level scope when using our primitives.

grwlf replied:
> That's what I said no? (constant) scalar values never end up as constvars, because they are embedded directly into the IR as literals.

Hmm, I think you said "constvars are only used for constant (i.e. non-tracer) array values" while I am saying that constants can be used for tracers. And it is exactly the case where we capture tensor variables from the outside scope. 

>  I have tested examples where variables are hoisted out to the top-level scope when using our primitives.

Can you share some examples where constants propagated through more than one level of Jaxpr? (I am worrying about possible misunderstanding as well)

dime10 replied:
> Hmm, I think you said "constvars are only used for constant (i.e. non-tracer) array values" while I am saying that constants can be used for tracers. And it is exactly the case where we capture tensor variables from the outside scope.

Thanks for clarifying! Look at the description in the [jax docs](https://jax.readthedocs.io/en/latest/jaxpr.html#constant-vars), I'm pretty sure `constvars` are only ever used for constant values in the program (not values that where originally tracers).

> And it is exactly the case where we capture tensor variables from the outside scope.

This was the second insight above, `constvars` has nothing to do with capturing variables from the outside scope (which I originally believed). Instead it is used in connection with hoisting variables up into an outer scope. I think this is reasonable well explained in the [jaxpr doc page](https://jax.readthedocs.io/en/latest/jaxpr.html#understanding-jaxprs), but you have to look for mentions of `constvars`.

> Can you share some examples where constants are propagated through more than one level of Jaxpr? (I am worrying about possible misunderstanding as well)

For example: (you can add more nesting with same effect)

```py
@qjit
def f(x: int):
    @for_loop(0, 5, 1)
    def g(i):
        y = jnp.array([0.0, 1.0])
        debug.print(y)
    return g()
f.jaxpr
```

In the jaxpr you can see the float array is hoisted out to `a:f64[2]` before the `;`, so this argument is stored in the `consts` field of the outermost jaxpr:
```
{ lambda a:f64[2]; b:i64[]. let
     = for_loop[
      apply_reverse_transform=False
      body_jaxpr={ lambda ; c:f64[2] d:i64[]. let
           = debug_print[memref=False] c
        in () }
      body_nconsts=1
    ] 0 5 1 a 0
  in () }
```

If `y = jnp.array(0.1)` is a scalar on the other hand, it is stored as a literal:
```
{ lambda ; a:i64[]. let
     = for_loop[
      apply_reverse_transform=False
      body_jaxpr={ lambda ; b:i64[]. let
           = debug_print[memref=False] 0.1
        in () }
      body_nconsts=0
    ] 0 5 1 0
  in () }
```

And if `y = i` is a dynamic value, then it just uses that jaxpr variable (wherever it's from):
```
{ lambda ; a:i64[]. let
     = for_loop[
      apply_reverse_transform=False
      body_jaxpr={ lambda ; b:i64[]. let
            = debug_print[memref=False] b
        in () }
      body_nconsts=0
    ] 0 5 1 0
  in () }
```

grwlf replied:
Thanks @dime10. Consider this example:

``` python
@qjit
def f(x: int, y:jax.core.ShapedArray([2], int)):
    @for_loop(0, 5, 1)
    def g(i):
        debug.print(y)
    return g()

f.jaxpr
```

``` result
{ lambda ; a:i64[] b:i64[2]. let
     = for_loop[
      apply_reverse_transform=False
      body_jaxpr={ lambda ; c:i64[2] d:i64[]. let
           = debug_print[memref=False] c
        in () }
      body_nconsts=1  # <--- [1]
    ] 0 5 1 b 0
  in () }
```
 
The `y` here is a regular argument, but note the Jaxpr line marked with `[1]`: The `body_ncosts` equal to one means that Jaxpr tracer originally created a constant that was moved to argument by us just before the primitive lowering. At the time of the tracing, this constant stored a tracer. (Note, that I'm running Catalyst 0b1335c428f1, i.e. I am not in this PR's branch).

grwlf replied:
Yes, it is our variable indeed. But it has value of one which means that 1) Jax tracer emitter body jaxpr with one Jax constvar 2) We converted it into a regular parameter and recorded this fact by increasing the variable.

When I wrote about variable capturing, I meant that if we refer outside-scope variables in a Python program, Jax emits Jaxpr program with constants. Additionally, Jax creates a datatype called ClosedJaxpr, containing the values for these constants, in this case - the outside-scope tracers. But this is not compatible with our StableHLO lowerer, which requires tracers to be passed as positional arguments only. So we convert ClosedJaxpr into a ClosedJaxpr-with-empty-constansts by calling the function named `convert_constvars_jaxpr`. The assertion you mentioned just checks that this conversion really happened - it is something that Jax does not do by default.

dime10 commented on frontend/catalyst/utils/jax_extras.py:
```diff
@@ -477,3 +576,372 @@ def make_jaxpr_f(*args, **kwargs):
+    num_implicit_inputs: int | None = None,
+) -> Tuple[List[TracerLike], OutputType]:
+    """Deduce the Jax ``out_type`` given input and ouputs abstract entities. By abstract entities
+    we mean either Jax tracers or Jaxpr variables."""
```
What exactly is a Jaxpr variable (what is its Python class?) as opposed to a Jax tracer?

grwlf replied:
Both are distinct explicitly defined entities. Here is the [Jaxpr variable](https://github.com/google/jax/blob/88a60b808c1f91260cc9e75b9aa2508aae5bc9f9/jax/_src/core.py#L278) and here is the [Jax tracer](https://github.com/google/jax/blob/88a60b808c1f91260cc9e75b9aa2508aae5bc9f9/jax/_src/interpreters/partial_eval.py#L1522).
I see it as follows. Jax understands programs in two languages: Jaxpr (their own internal language) and Python.  Jax need to operate on lists of arguments and results somehow. So they use lists of tracers (in the context of tracing) or lists of Jaxpr variables interchangebly. See also this [Jax comment](https://github.com/google/jax/blob/88a60b808c1f91260cc9e75b9aa2508aae5bc9f9/jax/_src/core.py#L1677).

dime10 commented on frontend/catalyst/utils/jax_extras.py:
```diff
@@ -477,3 +576,372 @@ def make_jaxpr_f(*args, **kwargs):
+    expansion_strategy: ExpansionStrategy,
+    num_implicit_inputs: int | None = None,
+) -> Tuple[List[TracerLike], OutputType]:
+    """Deduce the Jax ``out_type`` given input and ouputs abstract entities. By abstract entities
```
Because JAX code is quite difficult to read, I think it would be very helpful if a docstring like this would not just say that it is deducing the output type, but also what that means concretely (for example "the output type should be this in so-and-so situation" or "the output type is deduced from the input types by doing so-and-so", or whatever the idea the programmer had in mind when writing this function).

grwlf replied:
The output type is an explicitly defined core entity of Jax, so I am assuming that the reader knows what it is. See the definition [here](https://github.com/google/jax/blob/88a60b808c1f91260cc9e75b9aa2508aae5bc9f9/jax/_src/core.py#L1310). So, I don't think that this function is the best place for the descriptions like this. Should we add a separate documentation/glossary on Jax into our dev docs and add a link to it here?

dime10 replied:
My comment is not asking to define what an output type is, but to explain what this function does. The short description is "it deduces output types from input types", but that is not enough imo for a maintainer to understand what is going on (which is super important!).
When you wrote this function, you had an understanding of what the function should be doing, which you then translated into code. What I'm asking for is a small number of sentences that could help someone reading the code quickly grasp this "understanding" you had of "what the function should be doing".

grwlf replied:
I see, please check the new comment.

dime10 replied:
Thanks, the added comments are super helpful :)

dime10 commented on frontend/catalyst/jax_primitives.py:
```diff
@@ -1377,19 +1440,30 @@ def _for_loop_lowering(
         body_args[0] = from_elements_op.result
 
-        # recursively generate the mlir for the loop body
+        # Re-order arguments in accordance with jax dynamic API convensions
```
One-line comments like this are actually incredible helpful to quickly decipher the different steps when reading a function. I would encourage putting these wherever possible, especially the parts with low-level JAX code (which is hard to read).

dime10 reviewed (commented):
I left a few comments from a high level pass through the PR, but I would like to understand the primitive and lowering implementations better before final approval, which will require another more in-depth reading.

grwlf commented on frontend/catalyst/pennylane_extensions.py:
```diff
@@ -1235,33 +1338,78 @@ def _call_with_quantum_ctx(self, ctx):
+        out_tree = out_sigs[-1].out_tree()
+        all_consts = [s.out_consts() for s in out_sigs]
+        out_types = [s.out_type() for s in out_sigs]
+        # FIXME: Creating output tracers like this is not quite cocrect. We need to calculate the
+        # unified result type first and only then use it as the output type of this primitive.
+        out_type2 = out_types[-1]
```
This is the issue I discovered several days ago. It was always present, but we didn't hit it because I believe we don't support conditionals in user-level quantum transformations (according to Erick). Here is the [issue](https://github.com/PennyLaneAI/catalyst/issues/403) that I created.

dime10 replied:
Thanks!

grwlf commented on frontend/catalyst/pennylane_extensions.py:
```diff
@@ -1457,7 +1605,7 @@ def _decorator(true_fn: Callable):
 
 
-def for_loop(lower_bound, upper_bound, step):
+def for_loop(lower_bound, upper_bound, step, preserve_dimensions: bool = False):
```
I see this option as a consequence of the following facts: (1) In loop body programs, we need results to match arguments (2) We only do one-pass tracing (of Python program representing loop body). All we have is one particular set of arguments and we need to determine one particular set of results. This might be not enough to emit the correct program, so we need additional information which we can get either from options (or by doing second-pass of the tracing which I assume we don't want). So, by this switch I plan to cover the two equally-important cases so I see it as compromise. The cases are: to preserve all shared dimension variables in loops and to treat all dimension variables as distinct ones, even if they happened to be shared in a particular 0-iteration inputs that we have.

dime10 replied:
Following the discussion from our meeting I think we settled on marking the flag experimental (`experimental_` prefix) and setting the default behaviour to `True`.

Don't forget to also documentation for this option.

josh146 replied:
👍

grwlf replied:
Done.

grwlf commented on frontend/test/pytest/test_conditionals.py:
```diff
@@ -30,8 +30,10 @@ def test_basic_cond_to_jaxpr(self):
     c:i64[] = cond[
-      branch_jaxprs=[{ lambda ; a:i64[] b:i64[]. let c:i64[] = integer_pow[y=2] a in (c,) },
-                     { lambda ; a:i64[] b:i64[]. let c:i64[] = integer_pow[y=3] b in (c,) }]
+      branch_jaxprs=[{ lambda ; a:i64[] b_:i64[]. let c:i64[] = integer_pow[y=2] a in (c,) },
```
This is our constant-padding code:) See [here](https://github.com/PennyLaneAI/catalyst/blob/f03253d8e5ffeee5b91ba79514057d8d813bc874/frontend/catalyst/utils/jax_extras.py#L269). We modify Jaxpr programs representing the cond branches using a simple copy-all-together algorithm. We make them to have the similar sets of constants which seems to be one of the StableHLO requirements.

dime10 replied:
Thanks! 
> We make them to have the similar sets of constants which seems to be one of the StableHLO requirements.

What changed that made this padding necessary? My thinking is that if this was a requirement of StableHLO, it would have already been a requirement before the dynamic API no?

The docstring of the linked function is also a good example of code that could use improved documentation:
> """Align the constants of Jaxpr programs. Return the list of corresponding programs accepting
>      same constants."""

In this case, what does it mean to "align the constants of a Jaxpr program"? And almost more importantly, _why_ is this being done?

grwlf replied:
> if this was a requirement of StableHLO, it would have already been a requirement before the dynamic API no?

Yes, it was. We used a different function for this, which didn't support dynamic api (but it did support some other modes that we don't need). I implemented a version which does support dynamism.

>  what does it mean to "align the constants of a Jaxpr program"?  And almost more importantly, why is this being done?

Basically, if we have two programs with constants  `a` and `b`, we transform them into the programs `a, b_` and `a_, b` where constants with `_` suffix are unused. Yes, it was the requirement of StableHLO before this PR.

dime10 replied:
Could you elaborate on how these two programs are used? Are you saying where we used to have a single jaxpr, there are now two?

grwlf replied:
No, by programs I meant the programs encoding the conditional branches. Users may write arbitrary many branches, each with its own set of constants (and, potentially, inputs, but for conditional we have zero inputs - it does make our life easier). We need to unify the signatures: they need to be the same so our StableHLO lowering can work. We do this unification by adding all constants to all programs. This is what I tried to say in the previous post. In the end we have the same number of programs, equal to the number of conditional branches, with equal set of constants.

dime10 replied:
Got it, I think I get the picture now. The conditional requires that the inputs for each branch are the same, which we have always performed by using existing JAX functionality. Now you wrote your own function to perform the equivalent task, but with dynamic array support. In your own implementation, you marked the unused constants for a program with an underscore, which is the change visible in this diff? :)

grwlf replied:
Just not inputs (cond branches have no inputs!), but constants. Yes, I wrote my own implementation of the padding algorithm which takes the variable shapes of _constant outer tracers_ (mind the discussion about constants!) into account.

grwlf commented on frontend/catalyst/utils/jax_extras.py:
```diff
@@ -477,3 +576,372 @@ def make_jaxpr_f(*args, **kwargs):
+    """Primitive containing nested Jaxpr programs accepting and returning values with dynamic
+    shapes."""
+
+    def bind(self, *args, **params):
```
I am adding the extended comment, but could you please clarify what exactly do you mean by EvalTrace?

The `bind` handler is called by both `primitive.bind` during the Python tracing and `eval_jaxpr` during the Jaxpr interpretation/tracing.

dime10 replied:
The [EvalTrace](https://github.com/google/jax/blob/a74a78250bd864bb9057bab77efd28c262feb5dd/jax/_src/core.py#L929) is a JAXPR interpreter. I think it is the default interpreter of JAXPR and just evaluates (runs) the primitives. You can see it has this behaviour because the `process_primitive` function just calls the primitives `impl` function.
This interpreter is also explained in the [autodidax tutorial](https://jax.readthedocs.io/en/latest/autodidax.html#evaluation-interpreter).

dime10 replied:
Do my comments make sense @grwl? In that this `bind` acts fundamentally different than the `bind` in JAX, by processing the primitive directly instead of letting the trace process the primitive?

grwlf replied:
I see.. Yes, this is a very good question. I think I might have misused the `bind` API of Jax, given its original purpose seen from the docs. I believe that the cleaner approach would patch DynamicJaxprTracer's `process_primitive` rather than the primitive's bind method
I think that Jax team also have weakened their design compared to the one described in the documentation, I am still looking into this.

grwlf replied:
FYI: I have implemented a [draft trace class](https://github.com/PennyLaneAI/catalyst/blob/b365193abec730f4b1f22b3e878ecc1507945408/frontend/catalyst/utils/jax_extras.py#L954) named `DynamicJaxprTraceEx` supporting dynamic shapes. In order to use it, however, I needed to overload more Jax routines, because the default DynamicJaxprTrace is hardcoded in a number of places in the `partial_eval` module which we currently rely on. Unfortunately, some tests fail with C-segfault, so futher investigation is required.. The code is located in the `dynshape-jax-tracer` branch of the catalyst repo.

grwlf commented on frontend/test/pytest/test_conditionals.py:
```diff
@@ -200,6 +202,11 @@ def cond_fn():
         ):
             qjit(qml.qnode(qml.device(backend, wires=1))(circuit))
 
+        with pytest.raises(
+            TypeError, match="Conditional requires consistent return types across all branches"
+        ):
+            qjit(circuit)
```
I believe we should, because we actually have different implementations for classical and quantum tracing. See [here](https://github.com/PennyLaneAI/catalyst/blob/2c8ce077743dd586025c759250ada01a57759835/frontend/catalyst/pennylane_extensions.py#L1363) vs [here](https://github.com/PennyLaneAI/catalyst/blob/2c8ce077743dd586025c759250ada01a57759835/frontend/catalyst/pennylane_extensions.py#L1398)

dime10 replied:
Sounds good 👍

dime10 commented on frontend/catalyst/utils/jax_extras.py:
```diff
@@ -477,3 +576,421 @@ def make_jaxpr_f(*args, **kwargs):
+    containing dynamic dimensions."""
+
+    def bind(self, *args, **params):
+        """Bind the Jax primitive into a Jaxpr program. This method are called during both the
```
```suggestion
        """Bind the Jax primitive into a Jaxpr program. This method is called during both the
```

grwlf commented:
@dime10 Iet me recap the discussion we had recently. Please, edit this comment according your view. Should we add more tasks here?

The list of tasks for this PR includes the following additional items:
* Write a Readme file describing the proposed tracing architecture ([Draft](https://github.com/PennyLaneAI/catalyst/blob/dynshape-quantum-primitives/FRONTEND.md)).
* Split the `utils/jax_extras.py` into `jax_extras/tracing.py`, `jax_extras/lowering.py` and `jax_extras/patches.py`.
* Code refactoring (Sergei expects David to put ToDo marks highlighting the problems)
  - Find better names for some classes/variables
  - Write comments where required

rmoyard commented on doc/changelog.md:
```diff
@@ -2,6 +2,31 @@
 
 <h3>New features</h3>
 
+* Catalyst control-flow primitives are now compatible with Jax dynamic API. For-loops, while-loops
```
```suggestion
* Catalyst control-flow primitives are now compatible with dynamic arrays. For-loops, while-loops
```

rmoyard commented on doc/changelog.md:
````diff
@@ -2,6 +2,31 @@
+  array([1., 1., 1.])
+  ```
+
+* A mitigation dialect (MLIR) was added. It initially contains a Zero Noise Extrapolation (ZNE) operation,
````
To be removed, it is from another PR

grwlf replied:
Fixed, thanks!

rmoyard commented on frontend/catalyst/jax_primitives.py:
```diff
@@ -1320,65 +1363,87 @@ def _while_loop_lowering(
-
-    loop_operands = []
-    for p in (lower_bound, upper_bound, step):
+    # result_types = loop_carry_types[1:]
```
Do we want to add those assert back?

grwlf replied:
Done, thanks!

rmoyard commented on frontend/catalyst/jax_primitives.py:
```diff
@@ -1320,65 +1363,87 @@ def _while_loop_lowering(
+    lower_bound = iter_args_plus_consts[body_nconsts + nimplicit + 0]
+    upper_bound = iter_args_plus_consts[body_nconsts + nimplicit + 1]
+    step = iter_args_plus_consts[body_nconsts + nimplicit + 2]
+    loop_index = iter_args_plus_consts[body_nconsts + nimplicit + 3]
```
Is it backward compatible if dynamic type is disabled by the user?

grwlf replied:
Could you please clarify what is dynamic tape? (by default, everything is expected to be backward-compatible with statically-shaped tensors)

rmoyard commented on frontend/catalyst/jax_tracer.py:
```diff
@@ -142,29 +145,32 @@ def _shapes(ts):
-    return ClosedJaxpr(jaxpr2, consts)
+    ctx, jaxpr: ClosedJaxpr, consts, target_types: List[ShapedArray], num_implicit_outputs
+):
+    with_qreg = len(target_types) > 0 and isinstance(target_types[-1], AbstractQreg)
```
Can you add some documentation?

grwlf replied:
Done.

rmoyard commented on frontend/catalyst/jax_tracer.py:
```diff
@@ -327,6 +356,18 @@ def has_nested_tapes(op: Operation) -> bool:
 
 
+def trace_function(ctx, fun, *args, expansion_strategy, **kwargs):
+    """Trace Python function supporting variable dimensions in its arguments and/or results"""
```
Maybe add that it should not contain any quantum code, right?

grwlf replied:
Indeed, thanks. Done.

rmoyard commented on frontend/catalyst/pennylane_extensions.py:
```diff
@@ -1058,32 +1064,62 @@ class ForLoop(HybridOp):
+            qrp.actualize(),
+        ]
+
+        out_expanded_classical_tracers = expand_results(
```
What does the expand results do?

grwlf replied:
"Expanded" is a recurring term for list of arguments/results. `expanded_` in the name of the list means that the implicit shape arguments were already added to the beginning of this list. I am working on the description document where I have added a corresponding [section](https://github.com/PennyLaneAI/catalyst/blob/dynshape-quantum-primitives/FRONTEND.md#expandedcollapsed-arguments-or-results).

rmoyard commented on frontend/catalyst/pennylane_extensions.py:
```diff
@@ -1380,33 +1483,78 @@ def _call_with_quantum_ctx(self, ctx):
+        out_tree = out_sigs[-1].out_tree()
+        all_consts = [s.out_consts() for s in out_sigs]
+        out_types = [s.out_type() for s in out_sigs]
+        # FIXME: Creating output tracers like this is not quite cocrect. We need to calculate the
```
```suggestion
        # TODO: Creating output tracers like this is not quite cocrect. We need to calculate the
```

rmoyard commented on frontend/catalyst/pennylane_extensions.py:
```diff
@@ -1419,7 +1567,7 @@ def __call__(self):
             return self._call_with_quantum_ctx(ctx)
         elif mode == EvaluationMode.CLASSICAL_COMPILATION:
-            return self._call_with_classical_ctx()
+            return self._call_with_classical_ctx(ctx)
```
Why do we need to add the context when classical? In cond I don't think you actually use it

grwlf replied:
I think I actually use it [here](https://github.com/PennyLaneAI/catalyst/pull/370/files/4478f648015e4fbc8c6b4cd359e7c96cb2e892a4#diff-8754bab66a7845aa8e14eb1ab898192548b9ab878d595f57b9c6639930e2c6bfR1538), in the `trace_function` call. This `ctx` is our context which keeps lower-level Jax tracing stack objects. This PR attempts to switch our code to a more uniform way of handling the nested tracing, thus the new state variable. In the previous version we in relied on Jax modifying its contexts without letting us know, see [here](https://github.com/google/jax/blob/88a60b808c1f91260cc9e75b9aa2508aae5bc9f9/jax/_src/interpreters/partial_eval.py#L2189).

rmoyard commented on frontend/catalyst/utils/jax_extras.py:
```diff
@@ -221,75 +264,132 @@ def __init__(self, boxid: int, e: JaxprEqn):
+
+    # pylint: disable=too-many-nested-blocks
+    all_padded_constvars = []
+    for jaxpr in jaxprs:
```
There is a 4x times nested loop, is there anyway to simplify that?

grwlf replied:
Indeed, we might reduce the number of loops by 1, thanks! Here we make every conditional branch (a Jaxpr program) to contain all constants found in all conditional branches. Without optimizations, we would get $O(B^2 * C * D)$ complexity where $B$ - number of branches, $C$ - maximum number of constant tensors, $D$ - maximum number of tensor dimensions. But we can pre-calculate the dimensions parts.

rmoyard reviewed (commented):
Thanks for the PR @grwlf , I have some initial comments!

My main concern is that the frontend is getting even more complex, for a very specific case. Could we imagine an abstraction where we check if we are dealing with dynamic shaped arrays and switch between a set of original functions and new functions (dynamic)? If not I think we need to benchmark the potential regressions for non-dynamic shaped arrays and see if we accept the results.

grwlf commented:
@rmoyard Thank you for starting the review!  I am currently in progress of addressing your suggestions. Let me share my thoughts regarding the abstraction and benchmarks points you made. 

> My main concern is that the frontend is getting even more complex, for a very specific case. Could we imagine an abstraction where we check if we are dealing with dynamic shaped arrays and switch between a set of original functions and new functions (dynamic)?

The goal of this PR was actually to avoid this style if possible, because otherwise we would need to double the number of tests, strictly speaking. In contrast, this PR aims at maintaining the backward compatibility: all the old static-case tests trigger the same codebase with the new ones that also check dynamically-shaped cases. Notably, all tests pass at the time of the writing. In case of concerns, I would rather request additional test scenarios, so I can extend our test set.  

> If not I think we need to benchmark the potential regressions for non-dynamic shaped arrays and see if we accept the results.

This is a non-trivial question. Ideally, the goal of meeting the performance requirements should have be announced as early as possible. In this PR I tried to enable new functionality while minimizing the number of changes instead. Below are some thoughts  specifically on performance:

* I am pretty sure that the tracing performance is going to degrade if measured in absolute time units. I suspect that the most troublesome place in code in this sense is [this one](https://github.com/PennyLaneAI/catalyst/blob/27854aa9700378fe585e348a8e4fe036e861d847/frontend/catalyst/utils/jax_extras.py#L839). Here we apply a hack in order to keep relying on the core `trace.frame.to_jaxpr2` function of Jax, which is its main Jaxpr IR generator. A faster implementation would rewrite this function which would IMO add too many compatibility risks.
* Having said that, I also think that the asymptotic tracing time is going to remain the same. We will still have a single-pass Python tracer as before, all the control-flows (loops, conditionals) are still lowered as operations.

dime10 commented:
> UPD: @rmoyard to think about it, running a benchmark is still a good thing to do anyways, do you have an idea what exactly to measure here?

I think the capture component is likely the only relevant thing here, so from starting the tracing of a user function to the MLIR generation. Ideally one of our more complex test cases (something that generates a decent bit of IR), and comparing the MLIR generation time before and after this PR.

dime10 commented on FRONTEND.md:
```diff
@@ -0,0 +1,533 @@
+
+- The MyPy type of types is `Tuple[Tuple[AbstractValue,bool],...]`
+  [link](https://github.com/google/jax/blob/88a60b808c1f91260cc9e75b9aa2508aae5bc9f9/jax/_src/core.py#L1304)
+- The `bool` tuple elements represent “explicitness” of argument in a source Python program.
+  Explicit arguments explicitly appear in user-defined Python functions. Implicit arguments, in
+  contrast, are added by Jax to match Jaxpr typing requirements.
```
Are explicit and implicit your own terminology or does it come from JAX?

dime10 commented on FRONTEND.md:
```diff
@@ -0,0 +1,533 @@
+
+### Primitives and binding
+
+Primitives are like functions in Jaxpr. If the primitive has a body, then binding is associated with
+tracing of the program representing primitive’s body. The binding of such a primitive includes:
```
I think we briefly talked about this in another comment, but to my understanding _binding_ does not refer to the tracing itself, rather tracing is a _prerequisite_ of binding a primitive with a body.
Is binding then the process by which a primitive registers itself in the program representation? (if applicable, since there are contexts in which no program representation is generated --> maybe it would be more accurate to say the process by which a primitive applies itself to the interpreter stack?)

dime10 replied:
(Aside:) If I think about the conclusion from this, do you know if that means that whenever an operation is invoked, the whole interpreter stack is immediately applied to it, rather than applying the stack once the whole program is captured? 🤔  (I guess this would apply in a jit context)

grwlf replied:
I am not entirely sure what do you mean by applying a stack to a primitive, could you elaborate? I would say we apply a primitive to a stack and get an updated stack. Did you mean this?

Normally (in the classical case) we would pass a complete inner Jaxpr program to the primitive we want to bind, maybe as an Jax-IR data structure, but it would be a complete program already. That means that we have to perform the full tracing of the inner program (btw, we could benefit from caching here, I think. It might be helpful e.g. if we want to add several instances of the same loop). For the outer program, binding just creates new equation in its equation list.

For quantum tracing we do preserve all the stacks and record operations in an aside structure rather then bind immediately. But I though it is not directly related to the dynamic API problem, so I did not include these details into the document. Let me know, I can add this into the document as well.

I don't particulary like that we use different structures for quantum tracing and a different tracing method for classical tracing. I'd say that a better tracing library would still use only one queue structure in all modes (speaking of forking of Jax))

dime10 replied:
> I am not entirely sure what do you mean by applying a stack to a primitive, could you elaborate? I would say we apply a primitive to a stack and get an updated stack. Did you mean this?

I was referencing the stack of interpreters that is setup by decorators like `grad` etc., which transform primitives in a certain way. So by "applying the interpreter stack to a primitive" I meant applying the processing function of each interpreter in the stack to the current primitive.

There was actually some great information shared about this on slack, so we can close this comment.

dime10 commented on FRONTEND.md:
````diff
@@ -0,0 +1,533 @@
+{ lambda ; a:i64[]. let
+    b:i64[] = add a 1
+    c:f64[b] = broadcast_in_dim[broadcast_dimensions=() shape=(None,)] 1.0 b
+  in (b, c) }
+```
+
+The first variable `b:i64[]` is the implicit result, while the second one is the explicit one.
````
It's strange that JAX has no way of tracking this in the representation right? Or is it just a printer problem, and with the in-memory representation you can clearly distinguish between explicit and implicit arguments/results?

grwlf replied:
The consequence is - several Python programs now might correspond to a single Jaxpr program. Yes, I think it might be a disadvantage, depending on their goals. They keep the implicitness information in types during the tracing, but they discard it at some point.

dime10 reviewed (commented):
This is a fantastic document, great work @grwlf!

grwlf commented:
@dime10 @rmoyard, Do you have any further updates regarding this PR? I would be glad to answer questions/suggestions.

dime10 commented:
> Thanks for the PR @grwlf , I have some initial comments!
> 
> My main concern is that the frontend is getting even more complex, for a very specific case. Could we imagine an abstraction where we check if we are dealing with dynamic shaped arrays and switch between a set of original functions and new functions (dynamic)? If not I think we need to benchmark the potential regressions for non-dynamic shaped arrays and see if we accept the results.

As suggested I ran a benchmark comparing 9bdfe8b8125db037feb02cd29da9ac3f236c95a2 (current PR head) against 96a8d1eb8531937461a9449da5af3354a7322675 (last sync point in main) using the Grover benchmark in the repo for 23 qubits.

What I measured is the time to run the single statement `fun.get_mlir(params)` using `%%timeit -r50`. The results show no measurable difference (exceeding observed fluctuations) between the two commits of around 345ms.

I conclude that from a performance point of view this PR is unproblematic 👍

grwlf commented:
This PR is outdated, so I am closing it. The same changes are already merged by:
* https://github.com/PennyLaneAI/catalyst/pull/775
* https://github.com/PennyLaneAI/catalyst/pull/777
